probe-rs/src/probe/jlink/mod.rs

```diff
@@ -941,6 +945,55 @@ impl DAPAccess for JLink {
     }
 }
 
+impl SwoAccess for JLink {
+    fn enable_swo(&mut self, config: &SwoConfig) -> Result<(), ProbeRsError> {
+        let jlink = self.handle.get_mut().unwrap();
+        self.swo_baud = Some(config.baud());
+        jlink
+            .swo_start_uart(config.baud(), SWO_BUFFER_SIZE.into())
+            .map_err(|e| ProbeRsError::Probe(DebugProbeError::ArchitectureSpecific(Box::new(e))))?;
+        Ok(())
+    }
+
+    fn disable_swo(&mut self) -> Result<(), ProbeRsError> {
+        let jlink = self.handle.get_mut().unwrap();
+        self.swo_baud = None;
+        jlink
+            .swo_stop()
+            .map_err(|e| ProbeRsError::Probe(DebugProbeError::ArchitectureSpecific(Box::new(e))))?;
+        Ok(())
+    }
+
+    fn read_swo_timeout(&mut self, timeout: std::time::Duration) -> Result<Vec<u8>, ProbeRsError> {
+        const MULTIPLIER: u32 = 2;
+
+        let end = std::time::Instant::now() + timeout;
+        let mut buf = vec![0; SWO_BUFFER_SIZE.into()];
+
+        let bytes_per_sec = self.swo_baud.unwrap() / 8;
+        let buffers_per_sec = std::cmp::max(1, bytes_per_sec / (buf.len() as u32)) * MULTIPLIER;
+        let poll_interval =
+            std::time::Duration::from_micros(1_000_000 / u64::from(buffers_per_sec));
+
+        let jlink = self.handle.get_mut().unwrap();
+
+        let mut bytes = vec![];
+        loop {
+            let data = jlink.swo_read(&mut buf).map_err(|e| {
```

Comment:
As far as I can work out, when you call `swo_read()` here, it will ask the probe how much data is available and immediately return it; i.e. it does not block beyond the time required for the USB round-trip. The timeout in `read_swo_timeout` is a maximum time you can block, not a minimum time you must block.

Consequently while it's neat having all the extra work to keep polling the device until the whole timeout has elapsed, I don't think you need to do it here -- the call to `swo_read()` won't take more than the specified timeout, and even if it does, you can't help it (since you block on that call anyway). Once that call returns the first time, you're OK to immediately return that data, since you're within the specified timeout and you have some data or nothing. That's basically what stlink and cmsis-dap (in polled mode) do, they issue a single request for as much data as the probe can send, receive it, and then return it immediately. The timeout essentially doesn't come into effect in the polled case; it's mostly for streaming interfaces (like in cmsis-dap) where you can do a single blocking read on the USB endpoint to just accumulate as much data as possible within a timeout.

However, the code here is exactly what an application would need to write to poll at the right interval, accumulate data, and then process it every `timeout` duration. Maybe we can move it into a new default method on `SwoAccess` that applications could call?

Reply:
The points you mentioned are valid and I think one should continue working on that and also consider proposing changes to the [jaylink](https://github.com/jonas-schievink/jaylink/) API.

This PR is to bring basic SWO support for JLink probes, which it achieves and could already help current users of probe-rs. Maybe a way forward would be to merge this PR and open an issue with your suggestion to structure this a bit further until we have another PR that makes `SwoAccess` more powerful. For sure, this could reuse parts of the code of this PR and move it to places that are more appropriate.

Reply:
:+1:, it does seem like this PR will work fine as-is so we could definitely change how `swo_read_timeout` works later instead.

Reply:
Would you take care of creating an issue so that we don't lose track of it?